outlines/models/transformers.py

```diff
@@ -0,0 +1,101 @@
+import math
+from typing import TYPE_CHECKING, List, Optional, Tuple, Union
+
+import numpy as np
+from numpy.typing import NDArray
+
+from outlines.models.tokenizer import Tokenizer
+
+if TYPE_CHECKING:
+    from transformers import PreTrainedModel, PreTrainedTokenizer
+
+
+__all__ = ["transformers"]
+
+
+class Transformers:
+    """Represents a `transformers` model."""
+
+    def __init__(
+        self,
+        model: "PreTrainedModel",
+        tokenizer: "PreTrainedTokenizer",
+        device: Optional[str] = None,
+    ):
+        import torch
+
+        if device is None:
+            if torch.cuda.is_available():
+                self.device = "cuda"
+            elif torch.backends.mps.is_available():
+                self.device = "mps"
+            else:
+                self.device = "cpu"
+        else:
+            self.device = device
+
+        self.model = model.to(self.device)
+        self.tokenizer = tokenizer
+
+    def __call__(
+        self, input_ids: NDArray[np.int64], attention_mask: NDArray[np.int64]
+    ) -> NDArray[np.float64]:
+        import torch
+
+        # `transformers` model accept `input_ids` of size at most equal to 2. We
+        # thus reshape the input array, call the model and reshape the output
+        # logits.
+        batch_shape = input_ids.shape[:-1]
+        num_tokens = input_ids.shape[-1]
+        input_ids = input_ids.reshape(math.prod(batch_shape), num_tokens)
+
+        input_ids = torch.from_numpy(input_ids).to(self.device)
+        attention_mask = torch.from_numpy(attention_mask).to(self.device)
+
+        output = self.model(input_ids, attention_mask=attention_mask)
+
+        next_token_logits = output.logits[:, -1, :]
+        probs = torch.nn.functional.softmax(next_token_logits, dim=-1).squeeze()
+        probs = torch.atleast_2d(probs)
+        numpy_probs = probs.cpu().detach().numpy()
+
+        return numpy_probs.reshape(batch_shape + (-1,))
+
+
+class TransformersTokenizer(Tokenizer):
+    """Represents a tokenizer for models in the `transformers` library."""
+
+    def __init__(self, model_name: str):
+        from transformers import AutoTokenizer
+
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name, padding_side="left")
+        self.eos_token_id = self.tokenizer.eos_token_id
+
+        if not self.tokenizer.pad_token:
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+            self.pad_token_id = self.eos_token_id
+        else:
+            self.pad_token_id = self.tokenizer.pad_token_id
+            self.pad_token = self.tokenizer.pad_token
+
+    def encode(
+        self, prompt: Union[str, List[str]]
+    ) -> Tuple[NDArray[np.int64], NDArray[np.int64]]:
+        output = self.tokenizer(prompt, padding=True, return_tensors="np")
```

Comment:
Pass-through arguments here, too?

Reply:
We can use pass-through arguments here too, although it makes no sense to override the values of `padding` and `return_tensor`.